docs/design/bundle-state.md

```diff
@@ -1,68 +1,123 @@
-- You want to perform rolling updates via Deployment, etc., when the file is updated.
+- K8s Secrets
+  - Pros: limits access by namespace; Helm and Zarf's proven implementation
+  - Cons: hacking a secret resource for something isn't technically a secret
+- K8s ConfigMaps
+  - Pros: Easy to use
+  - Cons: Not as secure as secrets (any namespace can access)
+- K8s Custom Resources
+  - Pros: Custom resource designed to store bundle information
+  - Cons: Heavy-handed approach; we don't want to secret data to be easily manipulated by users
 
-Note:
-Use a Secret for sensitive data, which is similar to a ConfigMap but more secure.
+### Decision
 
-Use a custom resource (CRD or Aggregated API) if most of the following apply:
+We will use K8s Secrets to store UDS state. This decision was made because it is the most secure option and aligns with the way Helm and Zarf store state information.
 
-- You want to use Kubernetes client libraries and CLIs to create and update the new resource.
-- You want top-level support from kubectl; for example, kubectl get my-object object-name.
-- You want to build new automation that watches for updates on the new object, and then CRUD other objects, or vice versa.
-- You want to write automation that handles updates to the object.
-- You want to use Kubernetes API conventions like .spec, .status, and .metadata.
-- You want the object to be an abstraction over a collection of controlled resources, or a summarization of other resources.
+## State Contents and Location
 
+Each UDS bundle deployed in the cluster will have its own state secret. The state secret will take the following form:
 
-### Leverage package metadata
-Add bundle information to the package metadata.
-- There isn't a specific field we could use, but we could "hijack" an existing field like `description` and add relevant bundle information there.
+```go
+type PkgStatus struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Status  string `json:"status"`
+}
 
-Pros:
-- implementation would be pretty straightforward, we already do package manipulation in dev mode.
-- we would be able to leverage the `zarf package list` functionality to pull and filter on the bundle information that gets added.
+type BundleState struct {
+	Name        string      `json:"name"`
+	Version     string      `json:"version"`
+	PkgStatuses []PkgStatus `json:"packages"`
+	Status      string      `json:"status"`
+}
+```
 
-Cons:
-- we would be overloading the package metadata with bundle information, which could be confusing.
-- we compromise the integrity of the package by modifying it.
+The `BundleState` struct will be stored in the secret's `data` field as a base64 encoded JSON string.
 
-## Decision
-Use secrets to store metadata about a bundle that has been deployed to the cluster
+#### Namespace
+
+The UDS state secret will be stored in the `uds` namespace. If the `uds` namespace doesn't exist, the CLI will create it.
```

Comment:
Is there a security risk posed by using a shared namespace for all bundle state secrets? Also curious if this could lead to conflicting state secrets if the same bundle were deployed in more than one namespace.

Reply:
> Is there a security risk

IMO no, the state secrets don't contain any sensitive data

> conflicting state secrets if the same bundle were deployed in more than one namespace

Good point that we hadn't considered. I don't _think_ this will be an issue because bundles themselves aren't namespaced, thoughts?

Reply:
Fair point, I didn't consider that bundles aren't namespace scoped. Shouldn't be an issue then I don't think.